src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/CheckForTypeRename.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System.ComponentModel.Composition;
+using System.Globalization;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Rename
+{
+    [Export(typeof(IProjectTreeActionHandler))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    [Order(10)]
+    internal class CheckForTypeToRename : ProjectTreeActionHandlerBase
+    {
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private readonly IEnvironmentOptions _environmentOptions;
+        private readonly IUserNotificationServices _userNotificationServices;
+        private readonly IRenameTypeService _renameTypeService;
+
+        [ImportingConstructor]
+        public CheckForTypeToRename(IUnconfiguredProjectVsServices projectVsServices,
+                                  IEnvironmentOptions environmentOptions,
+                                  IUserNotificationServices userNotificationServices,
+                                  IRenameTypeService renameTypeService)
+        {
+            _projectVsServices = projectVsServices;
+            _environmentOptions = environmentOptions;
+            _userNotificationServices = userNotificationServices;
+            _renameTypeService = renameTypeService;
+        }
+
+        public override async Task RenameAsync(IProjectTreeActionHandlerContext context, IProjectTree node, string value)
+        {
+            if (!node.Flags.Contains(ProjectTreeFlags.SourceFile))
+            {
+                await base.RenameAsync(context, node, value);
+                return;
+            }
+
+            string filePath = node.FilePath;
+
+            // Check if there are any symbols that need to be renamed
```

Comment:
From this logic down, I would just move into the rename type service; call it PromptForRenameIfRequired or something similar - it feels like this is very heavily tied to the implementation of how the rename service is implemented. 

If for example, we enabled the ability to wait until the language-service had published a particular version that was up-to-date with the tree, I would expect that to be hidden behind a single service, but because below is heavily tied to implementation, we'd need to change below to stop the queuing logic.

Reply:
Assuming you need to check whether you can rename type before the actual file on disk is renamed, I'm thinking something:

```  C#
bool renameType = await _renameTypeService.CanRenameTypeAsync(filePath);

await base.RenameAsync(context, node, value);

if (renameType)
   await _renameTypeService.RenameTypeAsync(filePath); <!-- Shows thread await & waits for the language service to see rename and then does rename.
```